src/graphics/canvas.rs

```diff
@@ -538,8 +565,105 @@ impl<'a> Canvas<'a> {
         self.draw_mesh_instances(&self.rect_mesh.clone(), image, instances, z, skip_z)
     }
 
+    /// Draws a section text that is fit and aligned into a given `rect` bounds.
+    ///
+    /// The section can be made up of multiple [Text], letting the user have complex formatting
+    /// in the same section of text (e.g. bolding, highlighting, headers, etc).
+    ///
+    /// [TextLayout] determines how the text is aligned in `rect` and whether the text wraps or not.
+    ///
+    /// Depth must be enabled to draw text.
+    pub fn draw_bounded_text(&mut self, text: &[Text], rect: Rect, layout: TextLayout) {
+        assert!(self.depth.is_some());
+
+        let mut section = wgpu_glyph::Section::default()
+            .with_screen_position((rect.x, rect.y))
+            .with_bounds((rect.w, rect.h))
+            .with_layout(match layout {
+                TextLayout::SingleLine { h_align, v_align } => {
+                    wgpu_glyph::Layout::default_single_line()
+                        .h_align(h_align.into())
+                        .v_align(v_align.into())
+                }
+                TextLayout::Wrap { h_align, v_align } => wgpu_glyph::Layout::default_wrap()
+                    .h_align(h_align.into())
+                    .v_align(v_align.into()),
+            });
+
+        for text in text {
+            let z = text.z.unwrap_or_else(|| {
+                self.z_pos += Z_STEP;
+                self.z_pos
+            });
+
+            section = section.add_text(wgpu_glyph::Text {
+                text: text.text,
+                scale: text.size.into(),
+                font_id: *self.fonts.get(text.font).expect("invalid font name"),
+                extra: wgpu_glyph::Extra {
+                    color: text.color.into(),
+                    z,
+                },
+            });
+        }
+
+        self.glyph_brush.queue(section);
+    }
+
+    /// Unbounded version of `draw_bounded_text`.
+    pub fn draw_text(
+        &mut self,
+        text: &[Text],
+        pos: impl Into<mint::Vector2<f32>>,
+        layout: TextLayout,
+    ) {
+        let pos = pos.into();
+        self.draw_bounded_text(
+            text,
+            Rect::new(pos.x, pos.y, f32::INFINITY, f32::INFINITY),
+            layout,
+        )
+    }
+
     /// Finish drawing with this canvas.
-    pub fn finish(self) {}
+    #[allow(unsafe_code)]
+    pub fn finish(self) -> GameResult<()> {
+        if self.depth.is_some() {
+            let Canvas {
+                device,
+                staging_belt,
+                glyph_brush,
+                encoder,
+                target,
+                depth,
+                transform,
+                pass,
+                ..
+            } = self;
+
+            std::mem::drop(pass);
+
+            glyph_brush
+                .draw_queued_with_transform(
+                    device,
+                    staging_belt,
+                    unsafe { encoder.as_mut().unwrap() },
```

Comment:
Missing safety guarantee

Reply:
pass is the only thing borrowing encoder so that's explicitly dropped above
quite harmless anywho